### PR TITLE
i#4940: Fix the wrong aflags estimation in drreg

### PR DESCRIPTION
**Pull request by JerryYouxin**
Fix issue #4940 to make sure the functionality of `drreg_reserve_aflags`/`drreg_unreserve_aflags` and `drreg_aflags_are_dead` for conditional instructions like `cmovcc` and `sbb`.

### CODE REVIEW

File: ext/drreg/drreg.c
Comment (abhinav92003) [inline ext/drreg/drreg.c:831]: comment nit: I guess this comment can be summarised as `writing doesn't count if already read`, like the other comments here.

**Review by abhinav92003 — COMMENTED** [review]
Thank you for filing the issue and the PR. The CI tests shows some failures due to formatting. Please fix those and rerun.

File: ext/drreg/drreg.c
Comment (JerryYouxin) [reply ext/drreg/drreg.c]: Thanks! I have fixed the long comments with the summarized ones.

**Comment by abhinav92003** [comment]
It seems that this PR doesn't run our CI workflows automatically, because this is @JerryYouxin's first PR: 
```
After a contributor has at least one pull request merged into a project's repository, any future pull requests from that contributor's fork will automatically run workflows.
```
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks.

Someone with write access needs to approve these workflow runs everytime. I can do that when needed.

The last commit is only a comment fix, so not running this time though.

**Comment by JerryYouxin** [comment]
I have added the test similar to the [gist](https://gist.github.com/JerryYouxin/58fd5a24113f3eea69b4264919afda88), but I find that the test will fail with a segmentation fault in application code and I cannot figure out why.

**Comment by JerryYouxin** [comment]
@abhinav92003 I have implemented a minimal test for arithmetic flag liveness estimation in `drreg-test.dll.c` and `drreg-test.c`. I have tested that if this bug fix is not included, the test will fail with message `aflag liveness estimation of drx and drreg should always consist`. If my bug fix is included, the test will pass correctly.

File: suite/tests/client-interface/drreg-test.dll.c
Comment (abhinav92003) [inline suite/tests/client-interface/drreg-test.dll.c:158]: Are there any existing tests that hit this assert? If yes, then we can probably remove the new `test15` in drreg-test. If no, then please add a comment here that this is relevant to test15.

**Review by abhinav92003 — COMMENTED** [review]
Formatting is broken for newly added changes. Please see the recent clang-format CI run. You can install clang-format on your machine to automatically format your changes: https://dynamorio.org/page_code_style.html

File: suite/tests/client-interface/drreg-test.dll.c
Comment (JerryYouxin) [reply suite/tests/client-interface/drreg-test.dll.c:158]: I have tested without newly added test15 and the existing test can still hit this assertion when the fix is not included. I will delete the redundant test15 in the next commit.

Comment (JerryYouxin) [reply suite/tests/client-interface/drreg-test.dll.c]: The format is fixed using the latest `clang-format` tool.

File: suite/tests/client-interface/drreg-test.dll.c
Comment (JerryYouxin) [reply suite/tests/client-interface/drreg-test.dll.c]: Thanks! I have fixed this message in the latest commit.

**Comment by abhinav92003** [comment]
Also, please change the PR description (and final commit description when you merge this PR) according to the Commit Messages section here: https://dynamorio.org/page_code_reviews.html. Particularly, please add the Issue: #NNNNN and Fixes: #NNNNN` string. Also, I usually prefer the PR/commit title to be short (less than 60 chars) so that Github doesn't wrap it in some pages.

File: ext/drreg/drreg.c
Comment (abhinav92003) [reply ext/drreg/drreg.c]: Please mark comments as resolved when they are addressed. Marking this resolved now.

**Comment by JerryYouxin** [comment]
The check has passed in x86 but failed in ci-windows with vs2017. It seems that the bug may still exist in the windows platform.

```
133: Test command: D:\a\dynamorio\dynamorio\build_debug-internal-64\bin64\drrun.exe "-s" "90" "-quiet" "-debug" "-use_dll" "D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll" "-exit0" "-stderr_mask" "0xC" "-msgbox_mask" "0" "-dumpcore_mask" "0x7d" "-staged" "-code_api" "-c" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.dll.dll" "--" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.exe"
133: Test timeout computed to be: 600
133: drreg-test running
133: CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
134/259 Test #133: code_api|client.drreg-test ...................................***Failed  Required regular expression not found. Regex=[^drreg-test running
?
drreg-test finished
?
$
]  6.10 sec
drreg-test running
CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
```

**Comment by JerryYouxin** [comment]
@abhinav92003 I'm sorry that I cannot reproduce the test failure on my local windows machine, and I cannot figure out why the test is failed. Can you provide more details about the windows test (e.g., environment setup) to reproduce the failure for further analysis of the bug fixes?

**Comment by abhinav92003** [comment]
> @abhinav92003 I'm sorry that I cannot reproduce the test failure on my local windows machine, and I cannot figure out why the test is failed. Can you provide more details about the windows test (e.g., environment setup) to reproduce the failure for further analysis of the bug fixes?

We have seen some issues before where we weren't able to reproduce test failures locally. Fortunately, there's a way to ssh into the Github Actions machine where the failure happened to get more info. See the section on "Debugging Tests on Github Actions Runner" at https://dynamorio.org/page_test_suite.html. Essentially, you need to change the Github Actions workflow config in your branch to invoke `tmate` if a failure happened: like 9580af86bf6d09d7b4efb710d891f70ff54ba2c8, but for `dynamorio/.github/workflows/ci-windows.yml`.

When you have sshed into the GA runner, then you can use `-loglevel <n> -logdir <dir>` to get some logs (https://dynamorio.org/page_logging.html) or gdb too (you may need to install it there). See https://dynamorio.org/page_debugging.html also for other tips.

**Comment by abhinav92003** [comment]
@JerryYouxin We have granted the access required for you to run the CI workflows on your own. You should be able to run the workflows by pushing to this branch and also re-run a previously workflow from its page. This should help you try out the `tmate` debugging strategy that I mentioned in my previous comment.

**Comment by JerryYouxin** [comment]
> @JerryYouxin We have granted the access required for you to run the CI workflows on your own. You should be able to run the workflows by pushing to this branch and also re-run a previously workflow from its page. This should help you try out the `tmate` debugging strategy that I mentioned in my previous comment.

Unfortunately, I still cannot open the `tmate` session. And It seems that I still cannot run the workflow by myself or by pushing new commits.

**Comment by derekbruening** [comment]
> Unfortunately, I still cannot open the `tmate` session. And It seems that I still cannot run the workflow by myself or by pushing new commits.

You're listed as a Pending member of Committers: you didn't accept the invitation yet it looks like.

**Comment by abhinav92003** [comment]
> The check has passed in x86 but failed in ci-windows with vs2017. It seems that the bug may still exist in the windows platform.
> 
> ```
> 133: Test command: D:\a\dynamorio\dynamorio\build_debug-internal-64\bin64\drrun.exe "-s" "90" "-quiet" "-debug" "-use_dll" "D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll" "-exit0" "-stderr_mask" "0xC" "-msgbox_mask" "0" "-dumpcore_mask" "0x7d" "-staged" "-code_api" "-c" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.dll.dll" "--" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.exe"
> 133: Test timeout computed to be: 600
> 133: drreg-test running
> 133: CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
> 134/259 Test #133: code_api|client.drreg-test ...................................***Failed  Required regular expression not found. Regex=[^drreg-test running
> ?
> drreg-test finished
> ?
> $
> ]  6.10 sec
> drreg-test running
> CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
> ```

Were you able to figure out the root cause of this? Maybe try adding some print statements for debugging -- like printing the instr in each iteration of the new loop in `event_app_analysis` so that we know exactly which test case is failing the assert; and maybe also something in `drreg_are_aflags_dead` and `drx_aflags_are_dead` that can help figure out which aflags bits differ in liveness.

**Comment by JerryYouxin** [comment]
> > The check has passed in x86 but failed in ci-windows with vs2017. It seems that the bug may still exist in the windows platform.
> > ```
> > 133: Test command: D:\a\dynamorio\dynamorio\build_debug-internal-64\bin64\drrun.exe "-s" "90" "-quiet" "-debug" "-use_dll" "D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll" "-exit0" "-stderr_mask" "0xC" "-msgbox_mask" "0" "-dumpcore_mask" "0x7d" "-staged" "-code_api" "-c" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.dll.dll" "--" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.drreg-test.exe"
> > 133: Test timeout computed to be: 600
> > 133: drreg-test running
> > 133: CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
> > 134/259 Test #133: code_api|client.drreg-test ...................................***Failed  Required regular expression not found. Regex=[^drreg-test running
> > ?
> > drreg-test finished
> > ?
> > $
> > ]  6.10 sec
> > drreg-test running
> > CHECK failed ..\suite\tests\client-interface\drreg-test.dll.c:159: aflags liveness estimation of drx and drreg should always be consistent
> > ```
> 
> Were you able to figure out the root cause of this? Maybe try adding some print statements for debugging -- like printing the instr in each iteration of the new loop in `event_app_analysis` so that we know exactly which test case is failing the assert; and maybe also something in `drreg_are_aflags_dead` and `drx_aflags_are_dead` that can help figure out which aflags bits differ in liveness.

Sorry for my late reply. I found that the aflag check will fail on linux x86/64 platforms after merging the latest branch and I can now reproduce the failure on my local machine. In detail, the failure comes from the `sahf` instruction. This instruction will write 5 arithmetic flags (CF, PF, AF, ZF, and SF) except for the OF. Thus, the arithmetic flag liveness analysis of `drx` and `drreg` will not be consistent as `drx` utilizes `TESTALL` macro to check the arithmetic flags will be overwritten completely, whereas currently `drreg` only check the flags with `TESTANY` macro. 

```
DEAD FAILED (drreg 1, drx 0) INSTR: 0x00007fe006ea0b00
mov    %rax -> %gs:0x000000e0[8byte]	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
mov    %gs:0x000000d8[8byte] -> %rax	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
mov    $0x000000000000abcd -> %rax	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
sahf   %ah	IS_CTI=0, flags=f800, READS_ANY=0, WRITES_ANY=0
mov    $0x0000000000000001 -> %rbx	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
mov    $0x0000000000000002 -> %rbx	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
mov    $0x0000000000000003 -> %rbx	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
lahf    -> %ah	IS_CTI=0, flags=1f, READS_ANY=1, WRITES_ANY=0
mov    %rax -> %gs:0x000000d8[8byte]	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
mov    %gs:0x000000e0[8byte] -> %rax	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
cmp    %al $0x81	IS_CTI=0, flags=8f800, READS_ANY=0, WRITES_ANY=1
sahf   %ah	IS_CTI=0, flags=f800, READS_ANY=0, WRITES_ANY=0
mov    %gs:0x000000d8[8byte] -> %rax	IS_CTI=0, flags=0, READS_ANY=0, WRITES_ANY=0
jmp    $0x00007fe246dd1361	IS_CTI=1, flags=0, READS_ANY=0, WRITES_ANY=0

----------------------
CHECK failed /home/yx/dynamorio/suite/tests/client-interface/drreg-test.dll.c:371: aflags liveness estimation of drx and drreg should always be consistent
```

In this case, the `drreg` current implementation of `drreg_are_aflags_dead` may introduce bugs when the instrumented codes  overwrite the OF flag and the instruction followed by `sahf` will read the OF flag.

**Comment by JerryYouxin** [comment]
@abhinav92003  So I'm wondering that if I can change the implementation of `drx_aflags_are_dead` with `drreg_are_aflags_dead` internally without deleting the duplicated interfaces?

**Comment by abhinav92003** [comment]
> I found that the aflag check will fail on linux x86/64 platforms after merging the latest branch

Do you mean there's some new failure due to some recent master commit?

> Thus, the arithmetic flag liveness analysis of drx and drreg will not be consistent as drx simply utilizes TESTALL macro to check the arithmetic flags will be overwritten completely (and fails) and TESTANY macro to check if the arithmetic flags are loaded.

I see. So for the example in your comment drx incorrectly reports that aflags are live; and the reason for that is that it didn't mark the 5 bits (CF, PF, AF, ZF, and SF) dead when it saw the `sahf`. Perhaps we should track liveness for each aflags bit so that later reads of a dead bit doesn't count.

> So can I change the implementation of drx_aflags_are_dead with drreg_are_aflags_dead internally without deleting the duplicated interfaces?

If I understand correctly, we need a change in `drx_aflags_are_dead` only right? `drreg_are_aflags_dead` already does the right thing.

**Comment by abhinav92003** [comment]
> So can I change the implementation of drx_aflags_are_dead with drreg_are_aflags_dead internally without deleting the duplicated interfaces?

If your question is whether we want to continue having two duplicate methods that intend to do the same thing: drx could use drreg to find liveness. But I'm not sure if we want ext libraries depending on each other. @derekbruening  thoughts?

**Comment by abhinav92003** [comment]
>  But I'm not sure if we want ext libraries depending on each other. @derekbruening thoughts?

I found some precedents to that: drcovlib uses drx, and drbbdup uses drreg.

So I think it would make sense for drx to use drreg to determine aflags liveness. @derekbruening any reason we want to keep them separate? (not sure why they were so in the first place)

**Comment by derekbruening** [comment]
I didn't spend much time but from a glance I don't understand the bug description for drx_aflags_are_dead: if you're considering the 6 flags as a unit, `sahf` does not make them dead since it doesn't write `OF`.  Its simplification will treat flags as live when they're actually dead for a two-instruction combination of instructions that writes to all 6 flags across both of them or something like that; but generally that would not be a correctness issue in that direction.

drreg did not exist at the time that drx_aflags_are_dead() was added.

Later drx evolved a bit and these days drx already depends on drreg:
```
$ git grep 'use_DynamoRIO_extension(drx.*drreg'
ext/drx/CMakeLists.txt:use_DynamoRIO_extension(drx drreg)
ext/drx/CMakeLists.txt:use_DynamoRIO_extension(drx_static drreg_static)
```

**Comment by abhinav92003** [comment]
In short, `drx_aflags_are_dead` should return true when invoked at top of the following sequence. But it doesn't in the current implementation.
```
sahf
lahf
cmp _,_
```

This is because of the simplification employed by drx -- not treating each aflags bit separately.

> but generally that would not be a correctness issue in that direction.

 You mean overestimating aflags  as alive when they're really dead is not a correctness issue? Probably yes. But the correctness issue is in `drreg_aflags_liveness`. See 0fdb97fbe463ec478a88f53d12033331030802ec.

**Comment by JerryYouxin** [comment]
> drreg did not exist at the time that drx_aflags_are_dead() was added.
> 
> Later drx evolved a bit and these days drx already depends on drreg

As @derekbruening mentions, the drx already depends on drreg and `drx_aflags_are_dead` is aimed to do the same work as `drreg_aflags_are_dead`. So I think it is safe to make the implementation of `drx_aflags_are_dead` as a wrapper of `drreg_aflags_are_dead` to make the API consistent with the previous versions.

**Comment by derekbruening** [comment]
> > drreg did not exist at the time that drx_aflags_are_dead() was added.
> > Later drx evolved a bit and these days drx already depends on drreg
> 
> As @derekbruening mentions, the drx already depends on drreg and `drx_aflags_are_dead` is aimed to do the same work as `drreg_aflags_are_dead`. So I think it is safe to make the implementation of `drx_aflags_are_dead` as a wrapper of `drreg_aflags_are_dead` to make the API consistent with the previous versions.

SGTM.  We have precedence for this type of thing: drx_reserve_note_range() was moved to drmgr_reserve_note_range() and now the drx version is just a wrapper.

**Comment by JerryYouxin** [comment]
It seems that the tests for drreg/drx have passed successfully, but the test for `win32.xarch` has failed and currently I cannot figure out why.

**Comment by JerryYouxin** [comment]
```
8: Internal cmake changing into directory: D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/xarch
8: ======== CMake output     ======
8: The C compiler identification is MSVC 19.16.27045.0
8: Detecting C compiler ABI info
8: Detecting C compiler ABI info - done
8: Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe - skipped
8: Detecting C compile features
8: Detecting C compile features - done
8: The CXX compiler identification is MSVC 19.16.27045.0
8: Detecting CXX compiler ABI info
8: Detecting CXX compiler ABI info - done
8: Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/cl.exe - skipped
8: Detecting CXX compile features
8: Detecting CXX compile features - done
8: Building for x86
8: Version number: 8.0.18816
8: The ASM compiler identification is MSVC
8: Found assembler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x86/ml.exe
8: Found assembler: ml64.exe
8: Found message compiler: C:/Program Files (x86)/Windows Kits/10/bin/10.0.17763.0/x86/mc.exe
8: Found Perl: C:/Strawberry/perl/bin/perl.exe (found version "5.32.1") 
8: Performing Test cxx17_available
8: Performing Test cxx17_available - Failed
8: Targeting subsystem 5.01
8: Looking for MFC
8: Looking for MFC - found
8: Could NOT find ZLIB (missing: ZLIB_LIBRARY) (found version "1.2.11")
8: WARNING: vera++ not found: disabling code style checks
8: Using Visual Studio 12 generators
8: Performing Test no_pie_avail
8: Performing Test no_pie_avail - Failed
8: Processing tests and generating expected output patterns
8: Found C:/Strawberry/c/bin/libstdc++-6.dll, enabling api.symtest
8: Configuring done
8: Generating done
8: Build files have been written to: D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/xarch
8: ======== End CMake output ======
8: Change Dir: D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/xarch
8: 
8: Run Build Command(s):C:/projects/install/ninja/ninja.exe dynamorio && [1/105] Generating ../event_strings.h
8: [2/105] Generating ../events.h
8: MC: Compiling D:/a/dynamorio/dynamorio/core/win32/events.mc
8: [3/105] Building C object core\CMakeFiles\ntdll_imports.dir\win32\ntdll_imports.c.obj
8: [4/105] Building C object core\CMakeFiles\drlibc.dir\drlibc\drlibc.c.obj
8: [5/105] Building C object core\CMakeFiles\drlibc.dir\drlibc\drlibc_notdr_saferead.c.obj
8: [6/105] Building C object core\CMakeFiles\drlibc.dir\drlibc\drlibc_notdr_dcxt.c.obj
8: [7/105] Building C object core\CMakeFiles\drlibc.dir\drlibc\drlibc_notdr_printlog.c.obj
8: [8/105] Building ASM object core\CMakeFiles\drlibc.dir\drlibc\drlibc_xarch.asm.obj
8: FAILED: core/CMakeFiles/drlibc.dir/drlibc/drlibc_xarch.asm.obj 
8: cmd.exe /C "C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx86\x86\cl.exe /nologo -I..\..\..\..\core\drlibc -I..\..\..\..\core\ir\x86 -I..\..\..\..\core\arch\x86 -I..\..\..\..\core\win32 -I..\..\..\..\core\ir -I..\..\..\..\core\arch -I..\..\..\..\core\lib -I. -Iinclude\annotations -DDYNAMORIO_INTERNAL -DCPP2ASM -E /EP ..\..\..\..\core\drlibc\drlibc_xarch.asm > core\CMakeFiles\drlibc.dir\drlibc\drlibc_xarch.asm.obj.s && "C:\Program Files\CMake\bin\cmake.exe" -Dfile=core\CMakeFiles\drlibc.dir\drlibc\drlibc_xarch.asm.obj.s -P "D:/a/dynamorio/dynamorio/make/CMake_asm.cmake" && ml64.exe /nologo /Zi /Zd /c /Focore\CMakeFiles\drlibc.dir\drlibc\drlibc_xarch.asm.obj core\CMakeFiles\drlibc.dir\drlibc\drlibc_xarch.asm.obj.s"
8: drlibc_xarch.asm
8: 'ml64.exe' is not recognized as an internal or external command,
8: operable program or batch file.
8: [9/105] Building C object core\CMakeFiles\drlibc.dir\drlibc\drlibc_notdr_logfile.c.obj
8: [10/105] Building ASM object core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj
8: FAILED: core/CMakeFiles/drlibc.dir/drlibc/drlibc_x86.asm.obj 
8: cmd.exe /C "C:\PROGRA~2\MICROS~1\2017\ENTERP~1\VC\Tools\MSVC\1416~1.270\bin\Hostx86\x86\cl.exe /nologo -I..\..\..\..\core\drlibc -I..\..\..\..\core\ir\x86 -I..\..\..\..\core\arch\x86 -I..\..\..\..\core\win32 -I..\..\..\..\core\ir -I..\..\..\..\core\arch -I..\..\..\..\core\lib -I. -Iinclude\annotations -DDYNAMORIO_INTERNAL -DCPP2ASM -E /EP ..\..\..\..\core\drlibc\drlibc_x86.asm > core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s && "C:\Program Files\CMake\bin\cmake.exe" -Dfile=core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s -P "D:/a/dynamorio/dynamorio/make/CMake_asm.cmake" && ml64.exe /nologo /Zi /Zd /c /Focore\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s"
8: drlibc_x86.asm
8: 'ml64.exe' is not recognized as an internal or external command,
8: operable program or batch file.
8: ninja: build stopped: subcommand failed.
8: 
 14/259 Test   #8: win32.xarch ..................................................***Failed
```

It seems that the environment setup of the `win32.xarch` test on `vs2017-64` is corrupted.

**Comment by derekbruening** [comment]
> It seems that the environment setup of the win32.xarch test on vs2017-64 is corrupted.

Given that the test passes everywhere else and only fails in this PR, chances are the failure is due to a local change here.  I see this PR changes the .yml config and has lines requesting ml64, so I would think something there is the culprit.

**Comment by JerryYouxin** [comment]
@derekbruening Thanks! The previous environment error is from my changes specified for my local machine and the error is gone when I revert those commits. But the `AArch64 Precommit` check is still not triggered automatically and I don't know how to trigger this check manually.

**Comment by abhinav92003** [comment]
run arm tests

**Comment by abhinav92003** [comment]
For some reason my previous comment didn't invoke Jenkins. @AssadHashmi do we need to add @JerryYouxin to any ACLs?

**Comment by JerryYouxin** [comment]
The arm test is still not running. What should I do to run arm tests manually?

**Comment by abhinav92003** [comment]
> For some reason my previous comment didn't invoke Jenkins. @AssadHashmi do we need to add @JerryYouxin to any ACLs?

To verify if access control is indeed the issue, I created a PR from my forked DR repo: #5004. It does run the AArch64 test suite as expected. Let me check if I can add you to the whitelist on Jenkins myself. I see that other committers are there.

**Comment by abhinav92003** [comment]
run arm tests

**Comment by abhinav92003** [comment]
@JerryYouxin I added you to the AArch64 Precommit whitelist. For now I triggered the build manually, but in future it should trigger automatically when you push to a branch.

In the test run `code_api|tool.drcachesim.threads-with-config-file` failed. I don't think that's related to this PR; it's a known flaky test: #4954. I'll review the PR.

File: ext/drx/drx.c
Comment (abhinav92003) [inline ext/drx/drx.c:224]: Maybe just inline dr_get_current_drcontext here.

File: ext/drx/drx.c
Comment (abhinav92003) [inline ext/drx/drx.c:226]: The ASSERT can check `res == DRREG_SUCCESS` instead of the static `false`.

File: ext/drx/drx.c
Comment (abhinav92003) [inline ext/drx/drx.c:238]: Comparing this old implementation with `drreg_forward_analysis`: I see that the latter doesn't have this logic to continue on elision. It just bails on seeing a cti: https://github.com/DynamoRIO/dynamorio/blob/51e5f9c946c566e704e133c4b784f1547dfd3328/ext/drreg/drreg.c#L770

To me it seems a minor optimisation, and it should be fine to not have it anymore. But better to consult @derekbruening to confirm if this optimisation has any other significance.

**Review by abhinav92003 — COMMENTED** [review]
LGTM, pending the unresolved comments.

File: ext/drx/drx.c
Comment (derekbruening) [reply ext/drx/drx.c:238]: This is for -max_elide_{jmp,call} > 0 where DR makes a superblock by merging the target into the source block.  It's turned off for -code_api by default b/c most clients don't want it.

Comment (abhinav92003) [reply ext/drx/drx.c:238]: I guess then it should be safe for aflags liveness to not recognise such elision cases, as we don't expect `drx_aflags_are_dead` to be invoked if code manipulation apis are disabled.

File: ext/drx/drx.c
Comment (JerryYouxin) [reply ext/drx/drx.c]: Thanks! I have fixed in the latest commit.

File: ext/drx/drx.c
Comment (JerryYouxin) [reply ext/drx/drx.c]: Thanks! I have fixed in the latest commit.

File: ext/drx/drx.c
Comment (abhinav92003) [inline ext/drx/drx.c:223]: You can use `IF_DEBUG(res =)` to fix the unused var error, like elsewhere in drx.c too.

**Review by abhinav92003 — APPROVED** [review]
LGTM. Thanks for debugging and fixing the issue! Please take a look at the section on Final Commit Message before merging https://dynamorio.org/page_code_reviews.html.

File: ext/drx/drx.c
Comment (JerryYouxin) [reply ext/drx/drx.c]: Thanks! I replace DR_ASSERT_MSG with IF_DEBUG in the latest commit.

**Comment by abhinav92003** [comment]
You can use "run arm tests" in a separate comment to re-run any flaky CI test suite. If you find that there's no issue associated with some flaky test, please file one.

**Comment by abhinav92003** [comment]
> It seems that the `drreg-test` fails in AArch64 CI tests and I cannot access an available local AArch64 machine for further debugging.  @abhinav92003  Is there any way to debug this failing issue?

We're hitting this check: `CHECK failed /var/lib/jenkins/workspace/DynamoRIO-AArch64-Precommit/suite/tests/client-interface/drreg-test.dll.c:824: must use the freed up slot`

This is because the total number of spill slots available in drreg are now different than before, so drreg is making different decisions about which spill slots to use.

Note that the new call to `drx_init` requests additional slots from drreg internally. I think we can simply skip the `drx_init` and `drx_exit` calls. Do add a comment in `dr_init` in `drreg-test.dll.c` that:

We skip drx_init/drx_exit because that will change the number of total drreg spill slots available for the drreg tests. Skipping drx_init is okay because we use only drx_are_aflags_dead which doesn't need any drx state to be initialised.

**Comment by JerryYouxin** [comment]
I have fixed the `drreg-test` issue by skipping drx_init and drx_exit, but the `AArch64 Precommit` CI tests still fails in `drcache`.

```
====> FAILURE in release-external-64 <====
release-external-64: 47 tests passed, **** 4 tests failed, but ignoring 2 for i#2417: ****
	(ignore: i#2417) 	code_api|tool.drcachesim.phys 
	(ignore: i#2417) 	code_api|tool.drcacheoff.rseq 
	code_api|tool.drcachesim.threads-with-config-file 
	code_api|tool.drcachesim.coherence 
```

**Comment by abhinav92003** [comment]
run arm tests

File: suite/tests/client-interface/drreg-test.dll.c
Comment (abhinav92003) [inline suite/tests/client-interface/drreg-test.dll.c]: Please add the comment suggested on the PR's main thread, about why we need to skip drx_exit and drx_init calls.

**Comment by abhinav92003** [comment]
> Do add a comment in `dr_init` in `drreg-test.dll.c` that:
> 
> We skip drx_init/drx_exit because that will change the number of total drreg spill slots available for the drreg tests. Skipping drx_init is okay because we use only drx_are_aflags_dead which doesn't need any drx state to be initialised.

I think you forgot to add this comment.

> I have fixed the drreg-test issue by skipping drx_init and drx_exit, but the AArch64 Precommit CI tests still fails in drcache.

Both seem timeouts; try re-running by commenting "run arm tests". When I did that, only the `code_api|tool.drcachesim.threads-with-config-file` failure was left over, which is a known issue #4954. If it persists even after you re-run, we can consider adding it to the ignore list.

**Comment by JerryYouxin** [comment]
run arm tests

**Comment by JerryYouxin** [comment]
run arm tests

**Comment by abhinav92003** [comment]
Oh great, it passed this time. I have added a comment to #4954 anyway, noting the consecutive failures.

This PR seems good to go. Please modify the final commit message before merging as described here: https://dynamorio.org/page_code_reviews.html.

**Comment by derekbruening** [comment]
@JerryYouxin -- please avoid duplicating the title for future merges (the docs https://dynamorio.org/page_code_reviews.html#autotoc_md117 talk about the separate edit box for the title line):

```
commit 9822a6515e48446de727af3b58108d35d46ae431

    i#4940: Fix the wrong aflags estimation in drreg (#4941)
    
    i#4940: Fix the wrong aflags estimation in drreg (#4941)
    
    Fix the wrong aflags estimation in drreg to make sure the functionality of drreg_reserve_aflags/drreg_unreserve_aflags and drreg_aflags_are_dead for conditional instructions like cmovcc and sbb. Make drx_aflags_are_dead as a wrapper of drreg_aflags_are_dead.
    
    Fixes: #4940
```

**Pull request merged**
